services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/CommandRouterServiceImpl.java

```diff
@@ -237,20 +266,42 @@ private Duration getSanitizedLifespan(final Duration lifespan) {
 
     @Override
     public Future<CommandRouterResult> unregisterCommandConsumer(final String tenantId, final String deviceId,
-            final String adapterInstanceId, final Span span) {
+            final boolean sendEvent, final String adapterInstanceId, final Span span) {
 
         return deviceConnectionInfo
                 .removeCommandHandlingAdapterInstance(tenantId, deviceId, adapterInstanceId, span)
                 .recover(thr -> {
                     if (ServiceInvocationException.extractStatusCode(thr) != HttpURLConnection.HTTP_PRECON_FAILED) {
-                        LOG.info("error removing command handling adapter instance [tenant: {}, device: {}]", tenantId, deviceId, thr);
+                        LOG.info("error removing command handling adapter instance [tenant: {}, device: {}]", tenantId,
+                                deviceId, thr);
+                        // for no precon-failed errors continue and send the event
+                        return sendDisconnectEventIfNeeded(tenantId, deviceId, sendEvent, adapterInstanceId, span)
+                                // keep the original error and not propagate the event send result
+                                .recover(thr2 -> Future.failedFuture(thr)).compose(v -> Future.failedFuture(thr));
+                    } else {
+                        // for precon-failed errors do not send the event
+                        return Future.failedFuture(thr);
                     }
-                    return Future.failedFuture(thr);
                 })
+                .compose(v -> sendDisconnectEventIfNeeded(tenantId, deviceId, sendEvent, adapterInstanceId, span))
                 .map(v -> CommandRouterResult.from(HttpURLConnection.HTTP_NO_CONTENT))
                 .otherwise(t -> CommandRouterResult.from(ServiceInvocationException.extractStatusCode(t)));
     }
 
+    private Future<Void> sendDisconnectEventIfNeeded(final String tenantId, final String deviceId,
+            final boolean sendEvent, final String adapterInstanceId, final Span span) {
+        if (sendEvent) {
+            return tenantClient.get(tenantId, span.context()).compose(
+                    tenantObject -> sendDisconnectedTtdEvent(tenantObject, deviceId, adapterInstanceId,
+                            span.context()).onFailure(thr -> {
```

Comment:
Formatting: putting a line break before `compose` and `onFailure` would make this more readable.

Reply:
done